consul/snapshot_endpoint.go

```diff
@@ -0,0 +1,203 @@
+// The snapshot endpoint is a special non-RPC endpoint that supports streaming
+// for taking and restoring snapshots for disaster recovery. This gets wired
+// directly into Consul's stream handler, and a new TCP connection is made for
+// each request.
+//
+// This also includes a SnapshotRPC() function, which acts as a lightweight
+// client that knows the details of the stream protocol.
+package consul
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+
+	"github.com/hashicorp/consul/consul/snapshot"
+	"github.com/hashicorp/consul/consul/structs"
+	"github.com/hashicorp/go-msgpack/codec"
+)
+
+const (
+	// noSnapshotsInDevMode indicates that snapshots aren't available for a
+	// server in dev mode.
+	noSnapshotsInDevMode = "Snapshot operations not available in dev mode"
+)
+
+// dispatchSnapshotRequest takes an incoming request structure with possibly some
+// streaming data (for a restore) and returns possibly some streaming data (for
+// a snapshot save). We can't use the normal RPC mechanism in a streaming manner
+// like this, so we have to dispatch these by hand.
+func (s *Server) dispatchSnapshotRequest(args *structs.SnapshotRequest, in io.Reader,
+	reply *structs.SnapshotResponse) (io.ReadCloser, error) {
+
+	// Perform DC forwarding.
+	if dc := args.Datacenter; dc != s.config.Datacenter {
+		server, ok := s.getRemoteServer(dc)
+		if !ok {
+			return nil, structs.ErrNoDCPath
+		}
+		return SnapshotRPC(s.connPool, dc, server.Addr, args, in, reply)
+	}
+
+	// Perform leader forwarding if required.
+	if !args.AllowStale {
+		if isLeader, server := s.getLeader(); !isLeader {
+			if server == nil {
+				return nil, structs.ErrNoLeader
+			}
+			return SnapshotRPC(s.connPool, args.Datacenter, server.Addr, args, in, reply)
+		}
+	}
+
+	// Snapshots don't work in dev mode because we need Raft's snapshots to
+	// be readable. Better to present a clear error than one from deep down
+	// in the Raft snapshot store.
+	if s.config.DevMode {
```

Comment:
Would be nice if this was possible, as it lets you do some interesting things with bootstrapping vault on a dev cluster, then doing a save/restore into a PKI-enabled cluster. Anyways, room for future enhancement.